core/src/serve_repair.rs

```diff
@@ -170,6 +167,95 @@ struct ServeRepairStats {
+    staked_remaining: u64,
+    unstaked_remaining: u64,
+    staked_request: u64,
+    unstaked_request: u64,
+}
+
+impl RepairBudget {
+    const INTERVAL_MS: u64 = 1_000;
+    const MAX_BYTES_PER_INTERVAL: u64 = 12_000_000;
+    const MAX_STAKED_BYTES_PER_INTERVAL: u64 = 10_000_000;
+    const MAX_NODE_BYTES_PER_INTERVAL: u64 = 1_000_000;
+
+    fn new(bank_forks: &RwLock<BankForks>) -> Self {
+        let root_bank = bank_forks.read().unwrap().root_bank();
+        let epoch_schedule = root_bank.epoch_schedule();
+        let epoch = epoch_schedule.get_epoch(root_bank.slot());
+        let staked_nodes = root_bank.epoch_staked_nodes(epoch);
+        Self {
+            last_update: Instant::now(),
+            epoch_staked_nodes: staked_nodes,
+            node_used: HashMap::default(),
+            staked_remaining: 0,
+            unstaked_remaining: Self::MAX_BYTES_PER_INTERVAL,
+            staked_request: 0,
+            unstaked_request: 0,
+        }
+    }
+
+    fn take(&mut self, id: &Pubkey, bytes: u64) -> bool {
+        // Enforce a per-node cap on bandwith usage. Return early if the cap is
+        // hit to prevent the node from skewing the stake/unstaked capacity
+        // adjustments.
+        let node_used = self.node_used.entry(*id).or_default();
+        let requested_node_usage = node_used.saturating_add(bytes);
+        if requested_node_usage > Self::MAX_NODE_BYTES_PER_INTERVAL {
+            return false;
+        }
+        *node_used = requested_node_usage;
+
+        let staked = self
+            .epoch_staked_nodes
+            .as_ref()
+            .map(|staked_nodes| staked_nodes.contains_key(id))
+            .unwrap_or_default();
+        if staked {
+            self.staked_request = self.staked_request.saturating_add(bytes);
+            // staked nodes can consume unstaked bandwidth
+            if self.staked_remaining >= bytes {
+                self.staked_remaining -= bytes;
+                true
+            } else if self.unstaked_remaining >= bytes {
+                self.unstaked_remaining -= bytes;
+                true
+            } else {
+                false
+            }
+        } else {
+            self.unstaked_request = self.unstaked_request.saturating_add(bytes);
+            if self.unstaked_remaining >= bytes {
+                self.unstaked_remaining -= bytes;
+                true
+            } else {
+                false
+            }
+        }
+    }
+
+    fn update(&mut self, bank_forks: &RwLock<BankForks>) {
+        if self.last_update.elapsed() > Duration::from_millis(Self::INTERVAL_MS) {
+            let root_bank = bank_forks.read().unwrap().root_bank();
+            let epoch_schedule = root_bank.epoch_schedule();
+            let epoch = epoch_schedule.get_epoch(root_bank.slot());
+            let staked_nodes = root_bank.epoch_staked_nodes(epoch);
+            let staked_remaining = self.staked_request.min(Self::MAX_STAKED_BYTES_PER_INTERVAL);
```

Comment:
```rust
let staked_remaining = self.staked_request.min(Self::MAX_STAKED_BYTES_PER_INTERVAL);
```
What is the purpose of this line?

Reply:
This sets the staked bandwidth reserved for the interval based on the staked requests from the previous interval, capping the reserve to allow some remaining bandwidth for unstaked requests.

Reply:
Why does "the previous interval" matter?
We have these periods of time that repair is pretty low, but then it spikes due to some issues in the cluster, and then goes back down again. Why should the node allocate the repair budget during the spike based on low repair period before it and vice versa?